pkgs/by-name/sp/spotube/package.nix

```diff
@@ -1,63 +1,63 @@
 {
-  lib,
-  stdenv,
-  fetchurl,
-
   autoPatchelfHook,
   dpkg,
-  makeBinaryWrapper,
-  makeWrapper,
-  undmg,
-  wrapGAppsHook3,
-
+  fetchurl,
   glib-networking,
   gtk3,
+  lib,
   libappindicator,
   libnotify,
   libsoup_3,
+  makeBinaryWrapper,
+  makeWrapper,
   mpv-unwrapped,
-  xdg-user-dirs,
+  stdenv,
+  undmg,
   webkitgtk_4_1,
+  wrapGAppsHook3,
+  xdg-user-dirs,
 }:
 
 let
   pname = "spotube";
-  version = "3.8.3";
-
-  meta = {
-    description = "Open source, cross-platform Spotify client compatible across multiple platforms";
-    longDescription = ''
-      Spotube is an open source, cross-platform Spotify client compatible across
-      multiple platforms utilizing Spotify's data API and YouTube (or Piped.video or JioSaavn)
-      as an audio source, eliminating the need for Spotify Premium
-    '';
-    downloadPage = "https://github.com/KRTirtho/spotube/releases";
-    homepage = "https://spotube.krtirtho.dev/";
-    license = lib.licenses.bsdOriginal;
-    mainProgram = "spotube";
-    maintainers = with lib.maintainers; [ tomasajt ];
-    platforms = [
-      "x86_64-linux"
-      "x86_64-darwin"
-      "aarch64-darwin"
-    ];
-    sourceProvenance = with lib.sourceTypes; [ binaryNativeCode ];
-  };
-
-  fetchArtifact =
-    { filename, hash }:
-    fetchurl {
-      url = "https://github.com/KRTirtho/spotube/releases/download/v${version}/${filename}";
-      inherit hash;
-    };
+  version = "3.9.0";
+
+  src =
+    let
+      fetchArtifact =
+        { filename, hash }:
+        fetchurl {
+          url = "https://github.com/KRTirtho/spotube/releases/download/v${version}/${filename}";
+          inherit hash;
+        };
+      sources = {
+        "aarch64-linux" = fetchArtifact {
+          filename = "Spotube-linux-aarch64.deb";
+          hash = "sha256-KBuUAgUU6c/rBkkbpYjSarzckIoi+gRtCkumvtaoras=";
+        };
+        "x86_64-linux" = fetchArtifact {
+          filename = "Spotube-linux-x86_64.deb";
+          hash = "sha256-vzzK3csyKYP6fKKIoysziBsc2tqrg5LXS/6KoXBtNVI=";
+        };
+        "x86_64-darwin" = fetchArtifact {
+          filename = "Spotube-macos-universal.dmg";
+          hash = "sha256-wwIIKY+bmMJZigc2AK/QMg142uvZ+D6LOddzedJM5f8=";
+        };
+        "aarch64-darwin" = fetchArtifact {
+          filename = "Spotube-macos-universal.dmg";
+          hash = "sha256-wwIIKY+bmMJZigc2AK/QMg142uvZ+D6LOddzedJM5f8=";
+        };
```

Comment:
Personally, I think it's better to have the sources attrset contain the derivations instead of just the data.

This way we could expose `sources` as `passthru.sources` which would allow us to easily fetch and update all FOD hashes at the same time.

Example from my recent pending PR:
https://github.com/NixOS/nixpkgs/blob/645e841fa609a8df25fc9e269280044a24bfc054/pkgs/by-name/vo/voicevox-core/package.nix#L13-L44

In this case, we would have to do `passthru = { inherit sources; };` because we're using two different builders

Reply:
so is better keep on passthru like your code above or keep the way as my past commit?

Reply:
If you don't have a preference, please change it to this one.

Reply:
that's correct?

```
passthru.sources =
    let
      fetchArtifact =
        { filename, hash }:
        fetchurl {
          url = "https://github.com/KRTirtho/spotube/releases/download/v${version}/${filename}";
          inherit hash;
        };
    in
    {
      "aarch64-linux" = fetchArtifact {
        filename = "Spotube-linux-aarch64.deb";
        hash = "sha256-KBuUAgUU6c/rBkkbpYjSarzckIoi+gRtCkumvtaoras=";
      };
      "x86_64-linux" = fetchArtifact {
        filename = "Spotube-linux-x86_64.deb";
        hash = "sha256-vzzK3csyKYP6fKKIoysziBsc2tqrg5LXS/6KoXBtNVI=";
      };
      "x86_64-darwin" = fetchArtifact {
        filename = "Spotube-macos-universal.dmg";
        hash = "sha256-wwIIKY+bmMJZigc2AK/QMg142uvZ+D6LOddzedJM5f8=";
      };
      "aarch64-darwin" = fetchArtifact {
        filename = "Spotube-macos-universal.dmg";
        hash = "sha256-wwIIKY+bmMJZigc2AK/QMg142uvZ+D6LOddzedJM5f8=";
      };
    };
  src = finalAttrs.passthru.sources.${stdenv.hostPlatform.system};
  
  ```

Reply:
it should be something like this:
```nix
sources = ...
src = sources.${stdenv.hostPlatform.system};

linux = stdenv.mkDerivation {
  ...
  passthru = { inherit sources; }
  ...
};
darwin = stdenv.mkDerivation {
  ...
  passthru = { inherit sources; }
  ...
};

Reply:
```
sources =
    let
      fetchArtifact =
        { filename, hash }:
        fetchurl {
          url = "https://github.com/KRTirtho/spotube/releases/download/v${version}/${filename}";
          inherit hash;
        };
    in
    {
      "aarch64-linux" = fetchArtifact {
        filename = "Spotube-linux-aarch64.deb";
        hash = "sha256-KBuUAgUU6c/rBkkbpYjSarzckIoi+gRtCkumvtaoras=";
      };
      "x86_64-linux" = fetchArtifact {
        filename = "Spotube-linux-x86_64.deb";
        hash = "sha256-vzzK3csyKYP6fKKIoysziBsc2tqrg5LXS/6KoXBtNVI=";
      };
      "x86_64-darwin" = fetchArtifact {
        filename = "Spotube-macos-universal.dmg";
        hash = "sha256-wwIIKY+bmMJZigc2AK/QMg142uvZ+D6LOddzedJM5f8=";
      };
      "aarch64-darwin" = fetchArtifact {
        filename = "Spotube-macos-universal.dmg";
        hash = "sha256-wwIIKY+bmMJZigc2AK/QMg142uvZ+D6LOddzedJM5f8=";
      };
    };

  src = sources.${stdenv.hostPlatform.system};
```

```
linux = stdenv.mkDerivation {
  ...
  passthru = { inherit sources; }
  ...
};
darwin = stdenv.mkDerivation {
  ...
  passthru = { inherit sources; }
  ...
};
```

have some way to test this passthru before I push?

Reply:
`nix-build -A spotube.sources` should build all the sources

Reply:
```
[nix-shell:~/Downloads/nixpkgs]$ nix-build -A spotube.sources
these 2 derivations will be built:
  /nix/store/j4p94j0n8z6gj3wd17nk4fjf4i52hq3j-Spotube-linux-x86_64.deb.drv
  /nix/store/l24nksf52jb9m1fmffqfylhn061dk9lg-Spotube-linux-aarch64.deb.drv
building '/nix/store/l24nksf52jb9m1fmffqfylhn061dk9lg-Spotube-linux-aarch64.deb.drv'...
building '/nix/store/j4p94j0n8z6gj3wd17nk4fjf4i52hq3j-Spotube-linux-x86_64.deb.drv'...

trying https://github.com/KRTirtho/spotube/releases/download/v3.9.0/Spotube-linux-aarch64.deb
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed

trying https://github.com/KRTirtho/spotube/releases/download/v3.9.0/Spotube-linux-x86_64.deb
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
100 18.9M  100 18.9M    0     0  12.4M      0  0:00:01  0:00:01 --:--:-- 30.9M
100 19.2M  100 19.2M    0     0  10.6M      0  0:00:01  0:00:01 --:--:-- 20.8M
/nix/store/h4cacazy1imwddy70khln5gah2kljzlz-Spotube-macos-universal.dmg
/nix/store/9sk5igi898aiz6jzx9585yc445a8rfys-Spotube-linux-aarch64.deb
/nix/store/h4cacazy1imwddy70khln5gah2kljzlz-Spotube-macos-universal.dmg
/nix/store/wgsnv01ixhq0yl05zj1722f28v60y0d8-Spotube-linux-x86_64.deb
```

cool? sorry for asking but I never had use passthru.

Reply:
This is what was supposed to happen, so it's working.

`stdenv.mkDerivation { passthru.foo = "bar"; }` is esentially just `stdenv.mkDerivation { foo = "bar"; }`; but `foo` won't be converted into a bash variable. This is useful, because that conversion would evaluate the expression, which we might not want (in this case, we don't want all the sources to be passed to the builder).

Reply:
Okay, thanks for the explication, I'll use more it on futures pkgs.

More one question, is really necessary pass inherit meta on linux and Darwin?

```
darwin = stdenv.mkDerivation {
    inherit
      pname
      version
      meta
      src
      ;
}

linux = stdenv.mkDerivation {
    inherit
      pname
      version
      meta
      src
      ;
 }
```

Reply:
Yes, since it includes the description and other stuff needed.

Reply:
Because we are not using meta.description example inside of linux or Darwin.

Reply:
`meta` is just metadata, it's not supposed to be directly used by the package itself anywhere.
I hope that answers your question.

Reply:
Thank you, TIL !